### PR TITLE
Introduction of Joiners and Injectors

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
That's a big one...

* I've added two interfaces `SpanInjector` and `SpanJoiner`.
* `Tracer` extends those interfaces and I'm manually passing `SpanJoinerComposite` and `SpanInjectorComposite` to `DefaultTracer`. The two Composite classes have all `SpanInjector` and `SpanJoiner` interfaces injected. That way you have to just register your version of `SpanInjector` and `SpanJoiner` and it will be automatically seen by `Tracer`. 
    * The good side of this is that from the client's perspective it's enough to call the `Tracer's` method - the internals are hidden
    * We're aligned with OpenTracing on this directly
    * The problem (related to last point) with this approach is that `Tracer` CANNOT be injected to any of those two interfaces (`Injector` and `Joiner`) since there's going to be a circular dependency.
* I've added `SpanInjector` and `SpanJoiner` for Http filter, Feign, Zuul and RestTemplate. That way only a handful of production classes (other than implementations of those two interfaces) + tests use Span header names directly. Thanks to this it will be easier to do step 2 which is introduce configurable headers via properties.
* Removed the constraint that headers are set only if tracing is on (that there is a span in the current thread) for Zuul, Feign and Rest Template. Iv'e done it cause the constraint was set only for the additional Sleuth related headers - there was no consistency. 
    * Also due to technical reasons - I would have to (which I did in another branch) refactor the whole code to separate `SpanAccessor` from `Tracer`.
    * For that reason `Tracer` can't extend `SpanAccessor` cause if you want to use `SpanAccessor` in the `SpanInjector` and `SpanJoiner` you will have a circular dependency

@spencergibb @dsyer @adriancole  - WDYT?

Related to issue #19

### CODE REVIEW

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceHeaders.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ *     <li>{@code X-B3-ParentSpanId} 64 encoded bits</li>
+ *     <li>{@code X-B3-Sampled} Boolean (either “1” or “0”)</li>
+ * </ul>
+ *
+ * @author Marcin Grzejszczak
+ *
+ * @since 1.0.0
+ */
+@ConfigurationProperties("spring.sleuth.headers")
+public class TraceHeaders {
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceHeaders.java]: I'm not using this anywhere yet - you can disregard that class

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java [hunk 2]
```diff
@@ -62,10 +69,24 @@ public Sampler defaultTraceSampler() {
...
+			SpanNamer spanNamer, SpanLogger spanLogger,
+			SpanReporter spanReporter, List<SpanJoiner> spanJoiners,
+			List<SpanInjector> spanInjectors) {
 		return new DefaultTracer(sampler, random, spanNamer, spanLogger,
-				spanReporter);
+				spanReporter, new SpanJoinerComposite(spanJoiners),
+				new SpanInjectorComposite(spanInjectors));
+	}
+
+	// have to register a NoOp - otherwise autowiring won't work for trace manager
```
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java]: Not if it were moved to a static inner class and `@Autowired(required = false) List<SpanJoiner> spanJoiners = new ArrayList<>();` was used.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java [hunk 3]
```diff
@@ -29,24 +30,30 @@
...
-			return (Span) object;
+	private final TraceKeys traceKeys;
+
+	public MessagingSpanInjector(TraceKeys traceKeys) {
+		this.traceKeys = traceKeys;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T> void inject(Span span, T carrier) {
```
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java]: Why not put `<T>` on `SpanInjector`?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.messaging.Message;
+
+/**
+ * Creates a {@link SpanBuilder} from {@link Message}
+ *
+ * @author Marcin Grzejszczak
+ *
+ * @since 1.0.0
+ */
+public class MessagingSpanJoiner implements SpanJoiner {
```
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java]: Same for `SpanJoiner<T>`.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java (hunk 3 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java]: I even did that once :) But this is exactly how it is done in OpenTracing.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java (hunk 4 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java]: It's exactly how it is done in OpenTracing

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java (hunk 2 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java]: I've already tried that and it won't work cause Tracer is also a `SpanJoiner` so you will have a circular dependency cycle.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java (hunk 3 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java]: OK, what does that matter?

Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java]: Cause:

- we want to be compatible with OpenTracing 
- it doesn't change much cause still I have to do sth like "accept" from Observator pattern

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java (hunk 4 above)
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java]: Actually OT doesn't have this type, it has these methods on Tracer. When extracted into a type whose only concern is injection for a specific carrier, it makes sense to move the carrier type param to the type.

Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanJoiner.java]: Marcin think of this like gson or similar. Gson itself can serialize any
type, but to teach it a specific type, you use a jsonadapter which has a
type target parameter for the class being serialized.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java (hunk 3 above)
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java]: I think my other notes follow this, but this doesn't have to do with OT
unless the method is on the Tracer itself. The Tracer accepts variable
types as it is used in many contexts (ex messaging and also HTTP). This
class has a limited concern.

I would expect an implementation of OT to internally have inject adapters
for different carriers and the composite (tracer) would have a registry of
those adapters.

OT or not OT, methods are designed differently depending on the shape and
scope of the call site. It is a subtle, but important difference. I can
work with you on this today, and thanks for raising a straw man!

**Comment by adriancole** [comment]
summary of notes, based on some lengthy discussion :)

* bind generic type params to `Joiner` and `Injector`, similar to https://github.com/opentracing/opentracing-java/pull/11/files#diff-eb9c3460aba76aabc0de04b05e4a2b3d
* autowire `Joiner|Injector<X>` where the types are used. This makes it easier to understand.
* drop `Span*Composite` if this is a new feature. doing so lowers the scope of this PR.

Ex.
```java
// both are functional interfaces

// carrier needs to be mutable here. ex Message.Builder
interface Injector<T> {
  void inject(Span span, T carrier);
}

// carrier is not affected, so can be immutable. ex. Message
interface Joiner<T> {
  Span join(T carrier);
}
```

**Comment by marcingrzejszczak** [comment]
Ok I've refactored the code to have `SpanInjector` and `SpanExtractor`. I had gigantic issues with generics thus there's a couple of TODOs.

cc @dsyer @adriancole @spencergibb

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceHeaders.java (hunk 1 above)
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceHeaders.java]: probably PropagationHeaders

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java [hunk 5]
```diff
@@ -29,24 +30,26 @@
...
-		}
-		return null;
+	private final TraceKeys traceKeys;
+
+	public MessagingSpanInjector(TraceKeys traceKeys) {
+		this.traceKeys = traceKeys;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java]: weird that this needed a warning supression

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java [hunk 6]
```diff
@@ -55,27 +58,28 @@ public static Span getSpanFromHeader(Message<?> message) {
 	 * a tag set it will not get overridden.
 	 *
 	 * @param traceKeys - the global configuration for trace keys
-	 * @param message - message to which headers will be added
+	 * @param messageBuilder - message builder to which headers will be added
 	 * @param span - span from which headers will be taken
 	 * @return the input message with updated headers
 	 */
-	public static Message<?> addSpanHeaders(TraceKeys traceKeys, Message<?> message,
+	public void addSpanHeaders(TraceKeys traceKeys, MessageBuilder messageBuilder,
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java]: I'd take out the method dispatch as you'll notice the docs have drifted anyway

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/websocket/TraceWebSocketAutoConfiguration.java [hunk 7]
```diff
@@ -46,12 +56,28 @@ public void registerStompEndpoints(StompEndpointRegistry registry) {
...
 	}
 
 	@Override
 	public void configureClientInboundChannel(ChannelRegistration registration) {
 		registration.setInterceptors(
-				new TraceChannelInterceptor(this.tracer, this.traceKeys, this.random));
+				new TraceChannelInterceptor(this.tracer, this.traceKeys, this.spanExtractor, this.spanInjector));
+	}
+
+	// TODO: Qualifier + ConditionalOnProp cause autowiring generics doesn't work
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/websocket/TraceWebSocketAutoConfiguration.java]: thx for the TODOs

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/HttpServletResponseInjector.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ * @since 1.0.0
+ */
+class HttpServletResponseInjector implements SpanInjector<HttpServletResponse> {
+
+	@Override
+	public void inject(Span span, HttpServletResponse carrier) {
+		addResponseHeaders(span, carrier);
+	}
+
+	private void addResponseHeaders(Span span, HttpServletResponse response) {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/HttpServletResponseInjector.java]: I don't think dispatching here helps much.. how about inlining it?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java [hunk 9]
```diff
@@ -199,13 +153,32 @@ else if (skip) {
...
-		if (span != null) {
-			if (!response.containsHeader(Span.SPAN_ID_NAME)) {
-				response.addHeader(Span.SPAN_ID_NAME, Span.idToHex(span.getSpanId()));
-				response.addHeader(Span.TRACE_ID_NAME, Span.idToHex(span.getTraceId()));
+	/**
+	 * Creates a span and appends it as the current request's attribute
+	 */
+	private Span createSpan(HttpServletRequest request, HttpServletResponse response,
+			boolean skip, Span spanFromRequest, String name) {
+		if (spanFromRequest == null) {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: the level of indenting here makes it hard to read. I'd early return here

ex 

```java
if (spanFromRequest != null) return spanFromRequest;
// now everything can shift left one tab

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/AbstractTraceHttpRequestInterceptor.java [hunk 10]
```diff
@@ -34,38 +34,16 @@
...
-				this.tracer.isTracing()) {
-			request.getHeaders().add(name, value);
-		}
-	}
-
-	private void setIdHeader(HttpRequest request, String name, Long value) {
-		if (value != null) {
-			setHeader(request, name, Span.idToHex(value));
-		}
+	private void enrichWithTraceHeaders(Span span, HttpRequest request) {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/AbstractTraceHttpRequestInterceptor.java]: loose the dispatch and inline. that way, we call the same operation the same thing (injecting a span into a request as opposed to enriching with trace headers)

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/FeignRequestTemplateInjector.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+import feign.RequestTemplate;
+
+/**
+ * Span injector that injects tracing info to {@link RequestTemplate}
+ *
+ * @author Marcin Grzejszczak
+ *
+ * @since 1.0.0
+ */
+class FeignRequestTemplateInjector implements SpanInjector<RequestTemplate> {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/FeignRequestTemplateInjector.java]: I actually like this.. using the common term injector allows me to relate to all the different code easily. thanks

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/FeignResponseHeadersHolder.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.Map;
+
+/**
+ * Mutable holder for Feign Response headers
+ *
+ * @author Marcin Grzejszczak
+ *
+ * @since 1.0.0
+ */
+class FeignResponseHeadersHolder {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/FeignResponseHeadersHolder.java]: thanks for making this package private.. I'm thinking maybe we add a feign response builder at some point.

**Comment by adriancole** [comment]
thanks I know this was hard work. You've noted some follow-up concerns which can be done post merge

:shipit:

**Comment by marcingrzejszczak** [comment]
Done in https://github.com/spring-cloud/spring-cloud-sleuth/commit/7c5af8c84e7cdb4d821975feea966eb49cf7e2d8

**Pull request closed without merging**